internal/buffer/backup.go

```diff
@@ -14,11 +14,16 @@ import (
 	"github.com/zyedidia/micro/v2/internal/util"
 )
 
-const BackupMsg = `A backup was detected for this file. This likely means that micro
-crashed while editing this file, or another instance of micro is currently
-editing this file.
+const BackupMsg = `A backup was detected for:
 
-The backup was created on %s, and the file is
+%s
+
+This likely means that micro crashed while editing this file,
+or another instance of micro is currently editing this file,
+or the necessary permission to overwrite the file was not given,
```

Comment:
Hmm... Indeed, `overwriteFile()` and `os.WriteFile()` return error not only if it fails to write to the file (in the middle of writing) but also if it fails to even open the file. But this means that our implementation is broken and should be reworked: we keep the backup in the case when we have no good reason to keep it, since we know that the file is not corrupted at all (since we know that we even failed to open it for writing, so we had no chance to corrupt it).

(Just in case, I don't mean that we should just add a check whether the error is `fs.ErrPermission` or not. Open may fail for different reasons, not only because of lack of permissions.)

Reply:
> (Just in case, I don't mean that we should just add a check whether the error is `fs.ErrPermission` or not. Open may fail for different reasons, not only because of lack of permissions.)

I know, because we already added new checks like...
https://github.com/zyedidia/micro/pull/3273/files#diff-e33a9d01c51bf09d6ef92d1a430e36f49aefed071acf4668ec68a9b68200e996R244-R249
...but `os.Stat()` and its `FileInfo` isn't feasible enough.

So it looks like we need a "dry" try here. For `os.O_CREATE` it isn't really "dry", but with this we can at least check if we can create a new file in this directory first, since we're on the way to create and write it anyway.

Reply:
> Why???

Hm, valid question.
I confirm, that `util.SafeWrite()` is easy to fix and split up `os.WriteFile()` into its pieces.

For `overwriteFile()` I agree, that we've to split parts already available, but separated in its own logic. We could move the `util.OverwriteError` into `overwriteFile()` and check for `util.ErrOverwrite` in `safeWrite()` and assume that everything different from `util.ErrOverwrite` took place on open (and we remove the backup), but most probably this isn't the case in every scenario/propagated error.
We could also add a specific `util.ErrOpen`, to have a clear trigger when it took place, but this doesn't change the fact in case we identify the trigger inside `overwriteFile()` we already wrote a backup and thus unintentionally touched the FS, wasted an inode etc.pp.
So I've to assume this wasn't what you had in your mind, but splitting up `overwriteFile()` respective extracting the `os.OpenFile()` out of it (maybe forward the file into it) and execute it **before** we create the backup, as we now do in `util.SafeWrite()`.

Reply:
One further though regarding:
> Why???

What about the situation in which we open the file before we create the backup (since the open could fail and we don't want the backup then)...with `os.O_TRUNC`, but writing the backup fails? Usually we would close the target file now, but we opened with `os.O_TRUNC` and thus broke it without backup.
Don't we run in circles then?

Reply:
`O_TRUNC` is not essential either, we could open without `O_TRUNC`, and then `f.Truncate()`, right?

But I'm not sure why are you talking about opening the file before creating the backup (we always try to create the backup before we try to do anything write the target file, and we are not going to change that, right?)... You are talking about a lot of different things here, mostly unclear, so it's hard to follow what are you talking about at all.

Reply:
> `O_TRUNC` is not essential either, we could open without `O_TRUNC`, and then `f.Truncate()`, right?

Check.

> But I'm not sure why are you talking about opening the file before creating the backup (we always try to create the backup before we try to do anything with the target file, and we are not going to change that, right?)...

I'm puzzled right now. Wasn't exactly this the intention due to "[...] this means that our implementation is broken and should be reworked"?
As figured out...

> [...] since we know that we even failed to open it for writing, so we had no chance to corrupt it [...]

...we had no chance to brake anything, since we were not even able to open the file, but we already created the backup (see above) and thus causing the error message to appear etc.pp.
I don't like to create the backup before opening the target, realizing the open fails and thus removing the backup again, to prevent implausible messages.
Why even create the backup, when we're able to figure out that it doesn't make any sense?

So from my perspective it is the logical step to open (`O_WRONLY|O_CREATE`) the file before we create the backup. In case the open fails we don't create the backup. And all this by splitting up the relevant parts into:
1. `open()`
1.1. return on error
2. `defer close()`
3. `backup()`
3.1. return on error
4. `write()`
4.1. return on error
6. `sync()`

`overwriteFile()` is a bit more tricky...

Please correct me in case I'm totally wrong or misunderstood your comments.

Reply:
Ah indeed, good point. Well, that wasn't really the intention I had in my mind, since "creating the backup before opening the target, realizing the open fails and thus removing the backup again" is still not as bad as keeping the false-alarm backup, since at least it has no annoying user-visible effects [*]. 

In fact I'm not even sure it is so bad on its own. If it means less code complication (by keeping the simple scheme "first deal with the backup, then deal with the target file" and just extending the existing logic deciding whether to remove the backup or not), then maybe why not.

[*] Unless we fail to remove this unneeded backup, for example because we crashed before removing it. But that is a rather bad situation anyway, which could as well result in an actual corruption of the target file.

Reply:
I don't disagree. Let's see the actual code.